Add discord role syncronisation for commands

ashhhleyyy opened the pull request:
This introduces a soft dependency on player-roles(-api), along with new configuration options in the integrations config: discord_permission_map and discord_roles map:
```json5
{
  "integrations": {
    // *snip*

    // This specifies vanilla permission levels and can be used even when player-roles is not installed
    "discord_permission_map": {
      "123123123123123123": 4
    },
    // This specifies role names from player-roles. Every user is implicitly given the role discord_everyone as well to allow for base permission controls
    "discord_role_map": {
      "123123123123123123": "discord_core"
    }
  }
}
```
Both maps use the Discord IDs for the roles. This also depends on the new code pushed in NucleoidMC/nucleoid-backend@54e56a3

ashhhleyyy committed:
feat: Add role permission synchronisation for commands executed from Discord

This introduces a soft dependency on player-roles(-api), along with new configuration options in the integrations config: discord_permission_map and discord_roles map.

Gegy commented on build.gradle:
```diff
@@ -25,10 +25,13 @@ dependencies {
 	modCompileOnly "com.viaversion:viaversion-api:4.3.1"
+	modCompileOnly "dev.gegy:player-roles-api:1.5.2"
+	// Depend on the full mod to use internals that aren't in the API
+	modCompileOnly "dev.gegy:player-roles:1.5.2"
```
Maybe these should be in the API? 😄

ashhhleyyy replied:
Yeah some of the impls for things should probably be moved into the API, however extras (along with the rest of the libraries and minigames) are still on 1.18, but player-roles has already moved onto 1.19, and so I wasn't sure if this would then have to wait everything updates to be able to use the a newer player-roles version

haykam821 replied:
It seems like [`RoleOverrideMap`](https://github.com/NucleoidMC/player-roles/blob/aa6ab077e27165b03921d36f0f3a580d2e986ab5/src/main/java/dev/gegy/roles/override/RoleOverrideMap.java) could easily be replaced with a custom [`RoleOverrideReader`](https://github.com/NucleoidMC/player-roles/blob/aa6ab077e27165b03921d36f0f3a580d2e986ab5/api/src/main/java/dev/gegy/roles/api/override/RoleOverrideReader.java) implementation, since I don't believe the overrides need to be modified.

ashhhleyyy replied:
Ah I missed that a lot of the `RoleOverrideMap` methods have `default` impls, and as of the new commit it should now just be dependent on player-roles-api

Gegy commented on src/main/java/xyz/nucleoid/extras/integrations/relay/RemoteCommandIntegration.java:
```diff
@@ -35,7 +34,22 @@ public static void bind(NucleoidIntegrations integrations, IntegrationsConfig co
-                integration.commandQueue.add(new RemoteCommand(command, sender));
+                var discordRoles = body.get("roles").getAsJsonArray();
+                var roles = new ArrayList<String>();
+                roles.add("discord_everyone");
```
Nit: can we extract this out to a constant? (also, would it be reasonable to just be `everyone`?)

ashhhleyyy replied:
I didn't go with `everyone` because we probably don't want people on Discord to be able to do things like `/help` as it spams the channel and hits the send rate-limits because each line is a separate message

ashhhleyyy committed:
fix: Address things from review

This extracts "discord_everyone" into a constant and removes the dependency on the full player-roles mod by using a custom implementation of RoleOverrideReader
